src/backend/distributed/commands/multi_copy.c

```diff
@@ -2585,6 +2602,12 @@ CitusCopyDestReceiverShutdown(DestReceiver *destReceiver)
 	List *connectionStateList = ConnectionStateList(connectionStateHash);
 	FinishLocalCopy(copyDest);
 
+	/*
+	 * We are not going to need any reserved connections for this COPY command,
+	 * so proactively deallocate all the reserved connections right now.
+	 */
+	DeallocateAllReservedConnections();
```

Comment:
This call is bugging me a little.

It seems like it does more than it should, e.g. it decreases the reserved counter multiple times even if the COPY command only increased it once. That it is ok under the assumption that nothing else tries to reserve connection during the COPY (not strictly guaranteed because we might evaluate PL/pgSQL functions if there is a `a int default my_fn()` column), and that nothing else will leave behind reservations.

The other issue is that even after the COPY the transaction might still need additional connections to the workers for which we give up the reservations.

Is there any harm in not calling this function here at all and waiting until the end of the transaction?

Reply:
I added this for the cases like:
```
BEGIN;
COPY single_shard ..
COPY single_shard ..
...
```

Where, we'd keep reserving unnecessarily for the other nodes.

Considering the situation you described (which I'll  try to add a test), `DeallocateAllReservedConnections ()` is too aggressive here. It is good to have it for end of transactions, but we should probably have `DeallocateReservedConnections(numberOfDeallocation = 1)`.

And, for the transaction end,  we keep the existing behaviour?